// react-film/lib/FilmStrip.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _glamor = require("glamor");

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _CallFunction = _interopRequireDefault(require("./CallFunction"));

var _Context = _interopRequireDefault(require("./Context"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ROOT_CSS = (0, _glamor.css)({
  MsOverflowStyle: 'none',
  overflowX: 'scroll',
  overflowY: 'hidden',
  touchAction: 'manipulation',
  WebkitOverflowScrolling: 'touch',
  '&::-webkit-scrollbar': {
    display: 'none'
  },
  '& > ul': {
    display: 'flex',
    listStyleType: 'none',
    margin: 0,
    padding: 0
  }
});

var _default = function _default(props) {
  return _react.default.createElement(_Context.default.Consumer, null, function (context) {
    return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("div", {
      className: (0, _classnames.default)(ROOT_CSS + '', props.className),
      ref: context._setFilmStripRef
    }, _react.default.createElement("ul", null, _react.default.Children.map(props.children, function (child) {
      return _react.default.createElement("li", null, child);
    }))), _react.default.createElement(_CallFunction.default, {
      arg: _react.default.Children.count(props.children),
      fn: context._setNumItems
    }));
  });
};

exports.default = _default;
