// botframework-webchat-component/lib/SendBox/ConnectivityStatus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.connectConnectivityAlert = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _connectToWebChat = _interopRequireDefault(require("../connectToWebChat"));

var _Localize = require("../Localization/Localize");

var _ErrorNotificationIcon = _interopRequireDefault(require("../Attachment/Assets/ErrorNotificationIcon"));

var _WarningNotificationIcon = _interopRequireDefault(require("../Attachment/Assets/WarningNotificationIcon"));

var _classnames2 = _interopRequireDefault(require("classnames"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var connectConnectivityAlert = function connectConnectivityAlert() {
  for (var _len = arguments.length, selectors = new Array(_len), _key = 0; _key < _len; _key++) {
    selectors[_key] = arguments[_key];
  }

  return _connectToWebChat.default.apply(void 0, [function (_ref) {
    var connectivityStatus = _ref.connectivityStatus,
        language = _ref.language;
    return {
      connectivityStatus: connectivityStatus,
      language: language
    };
  }].concat(selectors));
};

exports.connectConnectivityAlert = connectConnectivityAlert;

var _default = connectConnectivityAlert(function (_ref2) {
  var styleSet = _ref2.styleSet;
  return {
    styleSet: styleSet
  };
})(function (_ref3) {
  var _classnames;

  var connectivityStatus = _ref3.connectivityStatus,
      language = _ref3.language,
      styleSet = _ref3.styleSet;
  return _react.default.createElement("div", {
    "aria-live": "polite",
    className: (0, _classnames2.default)((_classnames = {}, _defineProperty(_classnames, styleSet.errorNotification, connectivityStatus === 'error'), _defineProperty(_classnames, styleSet.warningNotification, connectivityStatus === 'connectingslow'), _classnames)),
    role: "status"
  }, connectivityStatus === 'connectingslow' ? _react.default.createElement(_react.default.Fragment, null, _react.default.createElement(_WarningNotificationIcon.default, null), (0, _Localize.localize)('SLOW_CONNECTION_NOTIFICATION', language)) : connectivityStatus === 'error' && _react.default.createElement(_react.default.Fragment, null, _react.default.createElement(_ErrorNotificationIcon.default, null), (0, _Localize.localize)('FAILED_CONNECTION_NOTIFICATION', language)));
});

exports.default = _default;
