// botframework-webchat-component/lib/SendBox/TextBox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.connectSendTextBox = exports.default = void 0;

var _glamor = require("glamor");

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _TypeFocusSink = require("../Utils/TypeFocusSink");

var _Localize = require("../Localization/Localize");

var _connectToWebChat = _interopRequireDefault(require("../connectToWebChat"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ROOT_CSS = (0, _glamor.css)({
  display: 'flex',
  '& > input': {
    flex: 1
  }
});

var connectSendTextBox = function connectSendTextBox() {
  for (var _len = arguments.length, selectors = new Array(_len), _key = 0; _key < _len; _key++) {
    selectors[_key] = arguments[_key];
  }

  return _connectToWebChat.default.apply(void 0, [function (_ref) {
    var disabled = _ref.disabled,
        language = _ref.language,
        scrollToEnd = _ref.scrollToEnd,
        sendBoxValue = _ref.sendBoxValue,
        setSendBox = _ref.setSendBox,
        submitSendBox = _ref.submitSendBox;
    return {
      disabled: disabled,
      language: language,
      onChange: function onChange(_ref2) {
        var value = _ref2.target.value;
        setSendBox(value);
      },
      onSubmit: function onSubmit(event) {
        event.preventDefault(); // Consider clearing the send box only after we received POST_ACTIVITY_PENDING
        // E.g. if the connection is bad, sending the message essentially do nothing but just clearing the send box

        if (sendBoxValue) {
          scrollToEnd();
          submitSendBox();
        }
      },
      value: sendBoxValue
    };
  }].concat(selectors));
};

exports.connectSendTextBox = connectSendTextBox;

var _default = connectSendTextBox(function (_ref3) {
  var styleSet = _ref3.styleSet;
  return {
    styleSet: styleSet
  };
})(function (_ref4) {
  var className = _ref4.className,
      disabled = _ref4.disabled,
      language = _ref4.language,
      onChange = _ref4.onChange,
      onSubmit = _ref4.onSubmit,
      styleSet = _ref4.styleSet,
      value = _ref4.value;
  var typeYourMessageString = (0, _Localize.localize)('Type your message', language);
  return _react.default.createElement("form", {
    className: (0, _classnames.default)(ROOT_CSS + '', styleSet.sendBoxTextBox + '', (className || '') + ''),
    onSubmit: onSubmit
  }, _react.default.createElement(_TypeFocusSink.Context.Consumer, null, function (_ref5) {
    var sendFocusRef = _ref5.sendFocusRef;
    return _react.default.createElement("input", {
      "aria-label": typeYourMessageString,
      "data-id": "webchat-sendbox-input",
      disabled: disabled,
      onChange: onChange,
      placeholder: typeYourMessageString,
      ref: sendFocusRef,
      type: "text",
      value: value
    });
  }));
});

exports.default = _default;
