// botframework-webchat-core/lib/sagas/connectSaga.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _callee;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _effects = require("redux-saga/effects");

var _jsonwebtoken = require("jsonwebtoken");

var _mathRandom = _interopRequireDefault(require("math-random"));

var _updateConnectionStatus = _interopRequireWildcard(require("../actions/updateConnectionStatus"));

var _createPromiseQueue = _interopRequireDefault(require("../createPromiseQueue"));

var _botframeworkDirectlinejs = require("botframework-directlinejs");

var _connect = require("../actions/connect");

var _disconnect = require("../actions/disconnect");

var _marked =
/*#__PURE__*/
_regenerator.default.mark(observeAndPutConnectionStatusUpdate),
    _marked2 =
/*#__PURE__*/
_regenerator.default.mark(connectSaga),
    _marked3 =
/*#__PURE__*/
_regenerator.default.mark(_callee);

var CONNECTING = _botframeworkDirectlinejs.ConnectionStatus.Connecting,
    ONLINE = _botframeworkDirectlinejs.ConnectionStatus.Online,
    EXPIRED_TOKEN = _botframeworkDirectlinejs.ConnectionStatus.ExpiredToken,
    FAILED_TO_CONNECT = _botframeworkDirectlinejs.ConnectionStatus.FailedToConnect,
    ENDED = _botframeworkDirectlinejs.ConnectionStatus.Ended;

function randomUserID() {
  return "r_".concat((0, _mathRandom.default)().toString(36).substr(2, 10));
}

function observeAndPutConnectionStatusUpdate(directLine) {
  var connectionStatusQueue, connectionStatusSubscription, connectionStatus;
  return _regenerator.default.wrap(function observeAndPutConnectionStatusUpdate$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          connectionStatusQueue = (0, _createPromiseQueue.default)();
          connectionStatusSubscription = directLine.connectionStatus$.subscribe({
            next: connectionStatusQueue.push
          });
          _context.prev = 2;

        case 3:
          _context.next = 5;
          return (0, _effects.call)(connectionStatusQueue.shift);

        case 5:
          connectionStatus = _context.sent;
          _context.next = 8;
          return (0, _effects.put)((0, _updateConnectionStatus.default)(connectionStatus));

        case 8:
          _context.next = 3;
          break;

        case 10:
          _context.prev = 10;
          connectionStatusSubscription.unsubscribe();
          return _context.finish(10);

        case 13:
        case "end":
          return _context.stop();
      }
    }
  }, _marked, this, [[2,, 10, 13]]);
}

function negativeUpdateConnectionStatusAction(_ref) {
  var payload = _ref.payload,
      type = _ref.type;

  if (type === _updateConnectionStatus.UPDATE_CONNECTION_STATUS) {
    var connectionStatus = payload.connectionStatus;
    return connectionStatus !== CONNECTING && connectionStatus !== ONLINE;
  }
}

function rectifyUserID(directLine, userIDFromAction) {
  var token = directLine.token;

  var _ref2 = (0, _jsonwebtoken.decode)(token) || {},
      userIDFromToken = _ref2.user;

  if (userIDFromToken) {
    if (userIDFromAction && userIDFromAction !== userIDFromToken) {
      console.warn('Web Chat: user ID is both specified in the Direct Line token and passed in, will use the user ID from the token.');
    }

    return userIDFromToken;
  } else if (userIDFromAction) {
    if (typeof userIDFromAction !== 'string') {
      console.warn('Web Chat: user ID must be a string.');
      return randomUserID();
    } else if (/^dl_/.test(userIDFromAction)) {
      console.warn('Web Chat: user ID prefixed with "dl_" is reserved and must be embedded into the Direct Line token to prevent forgery.');
      return randomUserID();
    }
  } else {
    return randomUserID();
  }

  return userIDFromAction;
}

function connectSaga(directLine) {
  var activitySubscription, _ref3, connectionStatus;

  return _regenerator.default.wrap(function connectSaga$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          // DirectLineJS start the connection only after the first subscriber for activity$, but not connectionStatus$
          activitySubscription = directLine.activity$.subscribe({
            next: function next() {
              return 0;
            }
          });
          _context2.prev = 1;

        case 2:
          _context2.next = 4;
          return (0, _effects.take)(_updateConnectionStatus.UPDATE_CONNECTION_STATUS);

        case 4:
          _ref3 = _context2.sent;
          connectionStatus = _ref3.payload.connectionStatus;

          if (!(connectionStatus === ONLINE)) {
            _context2.next = 10;
            break;
          }

          return _context2.abrupt("return", function () {
            activitySubscription.unsubscribe();
            directLine.end();
          });

        case 10:
          if (!(connectionStatus === ENDED || connectionStatus === EXPIRED_TOKEN || connectionStatus === FAILED_TO_CONNECT)) {
            _context2.next = 12;
            break;
          }

          throw new Error('Failed to connect');

        case 12:
          _context2.next = 2;
          break;

        case 14:
          _context2.prev = 14;
          _context2.next = 17;
          return (0, _effects.cancelled)();

        case 17:
          if (!_context2.sent) {
            _context2.next = 20;
            break;
          }

          activitySubscription.unsubscribe();
          throw new Error('Cancelled');

        case 20:
          return _context2.finish(14);

        case 21:
        case "end":
          return _context2.stop();
      }
    }
  }, _marked2, this, [[1,, 14, 21]]);
}

function _callee() {
  var _ref4, _ref4$payload, directLine, userIDFromAction, username, updateConnectionStatusTask, meta, endDirectLine, terminateAction;

  return _regenerator.default.wrap(function _callee$(_context3) {
    while (1) {
      switch (_context3.prev = _context3.next) {
        case 0:
          _context3.next = 2;
          return (0, _effects.take)(_connect.CONNECT);

        case 2:
          _ref4 = _context3.sent;
          _ref4$payload = _ref4.payload;
          directLine = _ref4$payload.directLine;
          userIDFromAction = _ref4$payload.userID;
          username = _ref4$payload.username;
          _context3.next = 9;
          return (0, _effects.fork)(observeAndPutConnectionStatusUpdate, directLine);

        case 9:
          updateConnectionStatusTask = _context3.sent;
          _context3.prev = 10;
          meta = {
            userID: rectifyUserID(directLine, userIDFromAction),
            username: username
          };
          endDirectLine = void 0;
          _context3.next = 15;
          return (0, _effects.put)({
            type: _connect.CONNECT_PENDING,
            meta: meta
          });

        case 15:
          _context3.prev = 15;
          _context3.next = 18;
          return (0, _effects.call)(connectSaga, directLine);

        case 18:
          endDirectLine = _context3.sent;
          _context3.next = 26;
          break;

        case 21:
          _context3.prev = 21;
          _context3.t0 = _context3["catch"](15);
          _context3.next = 25;
          return (0, _effects.put)({
            type: _connect.CONNECT_REJECTED,
            error: true,
            meta: meta,
            payload: _context3.t0
          });

        case 25:
          return _context3.abrupt("continue", 49);

        case 26:
          _context3.prev = 26;
          _context3.next = 29;
          return (0, _effects.put)({
            type: _connect.CONNECT_FULFILLING,
            meta: meta,
            payload: {
              directLine: directLine
            }
          });

        case 29:
          _context3.next = 31;
          return (0, _effects.put)({
            type: _connect.CONNECT_FULFILLED,
            meta: meta,
            payload: {
              directLine: directLine
            }
          });

        case 31:
          _context3.next = 33;
          return (0, _effects.take)([_disconnect.DISCONNECT, negativeUpdateConnectionStatusAction]);

        case 33:
          terminateAction = _context3.sent;
          _context3.next = 36;
          return (0, _effects.put)({
            type: _disconnect.DISCONNECT_PENDING
          });

        case 36:
          endDirectLine();

          if (!(terminateAction.type === _disconnect.DISCONNECT)) {
            _context3.next = 40;
            break;
          }

          _context3.next = 40;
          return (0, _effects.take)(negativeUpdateConnectionStatusAction);

        case 40:
          _context3.prev = 40;
          _context3.next = 43;
          return (0, _effects.put)({
            type: _disconnect.DISCONNECT_FULFILLED
          });

        case 43:
          endDirectLine();
          return _context3.finish(40);

        case 45:
          _context3.prev = 45;
          _context3.next = 48;
          return (0, _effects.cancel)(updateConnectionStatusTask);

        case 48:
          return _context3.finish(45);

        case 49:
          _context3.next = 0;
          break;

        case 51:
        case "end":
          return _context3.stop();
      }
    }
  }, _marked3, this, [[10,, 45, 49], [15, 21], [26,, 40, 45]]);
}
